python/cuml/test/test_pickle.py

```diff
@@ -350,6 +366,58 @@ def test_unfit_pickle(model_name):
     assert mod_unpickled is not None
 
 
+def modified_clone(estimator, *, safe=True):
+    """
+    Identical to sklearn.clone except it uses equality comparison for numeric
+    types instead of "is"
+    """
+    estimator_type = type(estimator)
+    # XXX: not handling dictionaries
+    if estimator_type in (list, tuple, set, frozenset):
+        return estimator_type([clone(e, safe=safe) for e in estimator])
+    elif not hasattr(estimator, 'get_params') or isinstance(estimator, type):
+        if not safe:
+            return copy.deepcopy(estimator)
+        else:
+            if isinstance(estimator, type):
+                raise TypeError("Cannot clone object. " +
+                                "You should provide an instance of " +
+                                "scikit-learn estimator instead of a class.")
+            else:
+                raise TypeError("Cannot clone object '%s' (type %s): "
+                                "it does not seem to be a scikit-learn "
+                                "estimator as it does not implement a "
+                                "'get_params' method."
+                                % (repr(estimator), type(estimator)))
+
+    klass = estimator.__class__
+    new_object_params = estimator.get_params(deep=False)
+    for name, param in new_object_params.items():
+        new_object_params[name] = clone(param, safe=False)
+    new_object = klass(**new_object_params)
+    params_set = new_object.get_params(deep=False)
+
+    # quick sanity check of the parameters of the clone
+    for name in new_object_params:
+        param1 = new_object_params[name]
+        param2 = params_set[name]
+
+        # For simple values, use equality comparison. This is due to weird
+        # inconsistencies for `id()` with basic types such as float, int, str,
```

Comment:
I'm still not totally sure I understand why this is needed... if clone is going to fail but modified_clone works in one case, doesn't that mean users will still see a crash when they run regular clone? Probably I'm just misunderstanding

Reply:
This might be better suited for an offline discussion to cover the finer details but the main reason we can't use `clone` is because of this line: `param1 is not param2` (line 414). If you look at my psuedo-code (from a comment above) of what this test is doing, we need to compare the params of `new_model` and `old_model`. The `clone` function compares parameter values with the `is` keyword. Most of the time this works fine since the params just get passed around as references. However, this breaks down when crossing the cython/python boundary. Even basic types can get a new `id()`. For example,
```python
a = 5.0
estimator.a = a
assert a is estimator.a # Could fail if `a` is a property in a .pyx file
```

Instead, `modified_clone` is nearly identical to `clone` except it does value comparison (`==`) for basic types (`int`, `float`, `str`) and identity comparison for all others.

To your second point: I'm not sure the purpose of `clone` in production. It seems to me more like a testing utility function since it does unnecessary assertion over every parameter in both the new and old model. You could theoretically accomplish the same thing with `clone = lambda old_model: type(old_model)(**old_model.get_params())`

Reply:
I think this suggests a real issue with cloning our models, so I'm not sure the workaround (replacement clone is necessary here). Clone is used all over the place in dask-ml's HPO and model_selection functions in sklearn, which is why we added it originally. Are these tests newly failing that were previously passing (if you omit modified_clone)? Did the issue come up as more parameters were added to get_params?

Reply:
> I think this suggests a real issue with cloning our models

I'm not sure I agree. The models clone perfectly fine. It's the unnecessary asserts in `clone` that throw an issue.

> Are these tests newly failing that were previously passing (if you omit modified_clone)?

They are newly failing. They would pass before.

> Did the issue come up as more parameters were added to get_params?

Bingo. Most estimators returned very little or an incomplete set of param names which makes it easy to pass the test but functionally incorrect. Consider this example run in 0.16 branch:

```python
>>> old_model = cuml.DBSCAN(output_type="numpy")
>>> old_model.get_params()
{'eps': 0.5, 'min_samples': 5} # Missing 5 params
>>> new_model = clone(old_model) # No Assertion
>>> assert new_model.output_type == old_model.output_type
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```
> Clone is used all over the place in dask-ml's HPO and model_selection functions in sklearn, which is why we added it originally

This is good to know. Seems like we need some solution that works. Might need to get creative.

Reply:
Resolved this issue during internal discussion.